File: crates/ekore/src/harmonics/w1.rs
//! Harmonic sums of weight 1.
use num::complex::Complex;

use crate::harmonics::cache::{Cache, K};
use crate::harmonics::polygamma::cern_polygamma;

/// Compute the harmonic sum $S_1(N)$.
///
/// $$S_1(N) = \sum\limits_{j=1}^N \frac 1 j = \psi_0(N+1)+\gamma_E$$
/// with $\psi_0(N)$ the digamma function and $\gamma_E$ the Euler-Mascheroni constant.
pub fn S1(N: Complex<f64>) -> Complex<f64> {
    cern_polygamma(N + 1.0, 0) + 0.577_215_664_901_532_9
}

/// Analytic continuation of harmonic sum $S_{-1}(N)$ for even moments.
pub fn Sm1e(c: &mut Cache) -> Complex<f64> {
    let hS1 = c.get(K::S1);
    let hS1h = c.get(K::S1h);
    hS1h - hS1
}

/// Analytic continuation of harmonic sum $S_{-1}(N)$ for odd moments.
pub fn Sm1o(c: &mut Cache) -> Complex<f64> {
    let hS1 = c.get(K::S1);
    let hS1mh = c.get(K::S1mh);
    hS1mh - hS1
}

File: src/ekore/harmonics/w1.py
"""Weight 1 harmonic sums."""

import numba as nb
import numpy as np

from .polygamma import cern_polygamma


@nb.njit(cache=True)
def S1(N):
    r"""Compute the harmonic sum :math:`S_1(N)`.

    .. math::
      S_1(N) = \sum\limits_{j=1}^N \frac 1 j = \psi_0(N+1)+\gamma_E

    with :math:`\psi_0(N)` the digamma function and :math:`\gamma_E` the
    Euler-Mascheroni constant.

    Parameters
    ----------
    N : complex
        Mellin moment

    Returns
    -------
    S_1 : complex
        (simple) Harmonic sum :math:`S_1(N)`

    See Also
    --------
        ekore.harmonics.polygamma.cern_polygamma : :math:`\psi_k(N)`
    """
    return cern_polygamma(N + 1.0, 0) + np.euler_gamma


@nb.njit(cache=True)
def Sm1(N, hS1, hS1mh, hS1h, is_singlet=None):
    r"""Analytic continuation of harmonic sum :math:`S_{-1}(N)`.

    .. math::
      S_{-1}(N) = \sum\limits_{j=1}^N \frac {(-1)^j} j

    Parameters
    ----------
    N : complex
        Mellin moment
    hS1:  complex
        Harmonic sum :math:`S_{1}(N)`
    hS1mh: complex
        Harmonic sum :math:`S_{1}((N-1)/2)`
    hS1h: complex
        Harmonic sum :math:`S_{1}(N/2)`
    is_singlet: bool, None
        symmetry factor: True for singlet like quantities (:math:`\eta=(-1)^N = 1`),
        False for non-singlet like quantities (:math:`\eta=(-1)^N=-1`)

    Returns
    -------
    Sm1 : complex
        Harmonic sum :math:`S_{-1}(N)`

    See Also
    --------
        eko.anomalous_dimension.w1.S1 : :math:`S_1(N)`
    """
    if is_singlet is None:
        return (1 - (-1) ** N) / 2 * hS1mh + ((-1) ** N + 1) / 2 * hS1h - hS1
    if is_singlet:
        return hS1h - hS1
    return hS1mh - hS1
